src/print.c

```diff
@@ -245,6 +255,190 @@ launch_preview (const char *filename,
+                   int   w,
+                   int   h,
+                   int   page)
+{
+  GSubprocess *process;
+  GInputStream *stream;
+  unsigned char *data = NULL;
+  char *option1 = NULL;
+  char *option2 = NULL;
+  int read = 0;
+  gsize total_read = 0;
+
+  option1 = g_strconcat ("--file=", filename, NULL);
+  option2 = g_strdup_printf ("--raw=%d", page);
+  process = g_subprocess_new(G_SUBPROCESS_FLAGS_STDOUT_PIPE, NULL, "/usr/libexec/pdftoraw", option1, option2, NULL);
+
+  if (process)
+  {
+    stream = g_subprocess_get_stdout_pipe(process);
+    if (stream)
+    {
+      data = (unsigned char *) g_malloc( w * h * 4);
+      while ((read = g_input_stream_read(stream, data + total_read, 1024, NULL, NULL)) > 0)
+      {
+         total_read += read;
+      }
+    }
+  }
+  return data;
+}
+
+static void
+pdf_begin_print(GtkPrintOperation *op,
+                GtkPrintContext   *ctx,
+                char              *filename)
+{
+    (void)ctx;
+    gtk_print_operation_set_n_pages(op, pdf_get_actions_page(filename, PDF_NUM_PAGES, 0));
+}
+
+static void
+pdf_draw_page(GtkPrintOperation *op,
+              GtkPrintContext   *ctx,
+              gint               page_nr,
+              char              *filename)
+{
+    cairo_t *cr = NULL;
+    cairo_surface_t *surface = NULL;
+    int width = 0, height = 0;
+    unsigned char *data = NULL;
+    int stride = 0;
+    (void)op;
+
+    width = pdf_get_actions_page (filename, PDF_WIDTH, page_nr);
+    height = pdf_get_actions_page (filename, PDF_HEIGHT, page_nr);
+    cr = gtk_print_context_get_cairo_context(ctx);
+    stride = cairo_format_stride_for_width (CAIRO_FORMAT_ARGB32, width);
+    data = pdf_get_data_page (filename, width, height, page_nr);
+    if (data)
+    {
+      GdkPixbuf *pix = gdk_pixbuf_new_from_data (data, GDK_COLORSPACE_RGB, TRUE, 8, width, height, stride, NULL, g_free);
+      cairo_scale(cr, (72.0 / DPI), (72.0 / DPI));
+      gdk_cairo_set_source_pixbuf(cr, pix, 0, 0);
+      cairo_paint(cr);
+      cairo_fill (cr);
+      cairo_surface_destroy (surface);
+    }
+}
+
+static void
+pdf_end_print (GtkPrintOperation *operation,
+               GtkPrintContext   *context,
+               char              *filename)
+{
+  (void)operation;
+  (void)context;
+
+  if (filename)
+    unlink(filename);
+}
+
+static gboolean
+print_pdf(int                    fd,
+          XdpImplPrint         *object,
+          GDBusMethodInvocation *invocation)
+{
+  g_autofree char *filename = NULL;
+  g_autoptr(GUnixInputStream) istream = NULL;
+  g_autoptr(GUnixOutputStream) ostream = NULL;
+  GtkPrintSettings *settings = NULL;
+  GtkPrintOperation *print = NULL;
+  GError *err = NULL;
+  GVariantBuilder opt_builder;
+  int fd2;
+
+  istream = (GUnixInputStream *)g_unix_input_stream_new (fd, FALSE);
+
+  if ((fd2 = g_file_open_tmp (PACKAGE_NAME "XXXXXX", &filename, &err)) == -1)
+    return FALSE;
+
+  ostream = (GUnixOutputStream *)g_unix_output_stream_new (fd2, TRUE);
```

Comment:
Don't cast this, you never use the casted version anyway.

Reply:
Fixed by b351b31